### PR TITLE
Ask consent on SSO registration with default mxid

### LINKED ISSUE #10732: Consent flow skipped during SSO user registration if username is left at default
### Description

When using an SSO mapping provider that automatically sets a new user's mxid localpart (as opposed to prompting the user to choose it), then [requiring consent tracking on new user registration](https://github.com/matrix-org/synapse/blob/develop/docs/consent_tracking.md#enabling-consent-tracking-at-registration) doesn't work: registering a new user via SSO does not prompt for consent & lets the registration finish as if consent was not required.

### Steps to reproduce

- Follow the steps to [enable the user consent flow](https://github.com/matrix-org/synapse/blob/develop/docs/consent_tracking.md#collecting-policy-agreement-from-a-user) and to [make it mandatory on new user registration](https://github.com/matrix-org/synapse/blob/develop/docs/consent_tracking.md#enabling-consent-tracking-at-registration).
- Configure a SSO mapping provider that returns a mxid localpart
- Register a new user via SSO

<!--
Describe how what happens differs from what you expected.

If you can identify any relevant log snippets from _homeserver.log_, please include
those (please be careful to remove any personal or private data). Please surround them with
``` (three backticks, on a line on their own), so that they are formatted legibly.
-->

Expected results: The registration flow should eventually open a webpage at `/_matrix/consent` and not register a new user until the consent form on that page has been submitted.

Actual results: The registration will complete without prompting for consent, as if `user_consent.require_at_registration` was not set. Other consent configurations still take affect, though, such as `user_consent.block_events_error`.

### Version information

<!-- IMPORTANT: please answer the following questions, to help us narrow down the problem -->

<!-- Was this issue identified on matrix.org or another homeserver? -->
- **Homeserver**: Self-hosted

If not matrix.org:

<!--
 What version of Synapse is running?

You can find the Synapse version with this command:

$ curl http://localhost:8008/_synapse/admin/v1/server_version

(You may need to replace `localhost:8008` if Synapse is not configured to
listen on that port.)
-->
- **Version**: 1.41.1

- **Install method**: git clone
<!-- examples: package manager/git clone/pip  -->

- **Platform**: Fedora 34
<!--
Tell us about the environment in which your homeserver is operating
distro, hardware, if it's running in a vm/container, etc.
-->

### PR DESCRIPTION
**Pull request by AndrewFerr**
Fixes #10732

Signed-off-by: Andrew Ferrazzutti <fair@miscworks.net>

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
* [x] Pull request includes a [sign off](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#sign-off)
* [x] Code style is correct (run the [linters](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.md#code-style))

### CODE REVIEW

File: changelog.d/10733.bugfix [hunk 1]
```diff
@@ -0,0 +1 @@
+Allow user registration via SSO to require consent tracking for SSO mapping providers that don't prompt for Matrix ID selection.
```
Comment (richvdh) [inline changelog.d/10733.bugfix:1]: Feel free to add yourself here:

```suggestion
Allow user registration via SSO to require consent tracking for SSO mapping providers that don't prompt for Matrix ID selection. Contributed by @AndrewFerr.
```

**Review by richvdh — CHANGES_REQUESTED** [review]
Hi, thanks for the contribution!

> Note that I wasn't able to get the SyTest integration tests working, though unit tests & manual checks passed.

They run as part of our CI, so you're fine there.

> Regarding the change itself, a different approach than the one I used may be desired, as all I did was hack at a `UsernameMappingSession` to make it think that the user chose a username already, when it was really the mapping provider that did. This seems to do exactly what it needs to do, yet it still feels icky.

I sortof see what you mean, though I think broadly it is a sensible approach, if we consider it more as just going through the steps of the flow that are required. I think we might want to modify the implementation a bit though.

In particular, I'd suggest that we have a method which inspects an active `UsernameMappingSession` and redirects to the appropriate page for the next step. For example, if we don't yet have a `chosen_localpart`, we redirect to the username picker. If we have a `chosen_localpart` but need to gather consent, we redirect to the `new_user_consent` page. If we've got everything, we redirect to `/sso_register`. 

We could then call that method from `complete_sso_login_request`, `handle_submit_username_request` and `handle_terms_accepted`, which would reduce the amount of duplicated code.

Let me know what you think.

**Comment by AndrewFerr** [comment]
> I sortof see what you mean, though I think broadly it is a sensible approach, if we consider it more as just going through the steps of the flow that are required. I think we might want to modify the implementation a bit though.
> 
> In particular, I'd suggest that we have a method which inspects an active `UsernameMappingSession` and redirects to the appropriate page for the next step. For example, if we don't yet have a `chosen_localpart`, we redirect to the username picker. If we have a `chosen_localpart` but need to gather consent, we redirect to the `new_user_consent` page. If we've got everything, we redirect to `/sso_register`.
> 
> We could then call that method from `complete_sso_login_request`, `handle_submit_username_request` and `handle_terms_accepted`, which would reduce the amount of duplicated code.
> 
> Let me know what you think.

That sounds reasonable; thanks for the feedback. I can get something ready for it.

**Comment by AndrewFerr** [comment]
Okay--this should be ready now.

**Review by richvdh — APPROVED** [review]
looks great. Thank you so much!

**Pull request merged**